### PR TITLE
GH-1689: Implemented go to last edit location.

### LINKED ISSUE #1689: [editor-naviagtion] Support go to last edit location
It is a useful Eclipse feature. VSCode supports it via an extension: [`goto-last-edit-location`](https://marketplace.visualstudio.com/items?itemName=krizzdewizz.goto-last-edit-location). It is easy to support in Theia too.

Originally from here: https://gitter.im/theia-ide/theia?at=5ace98cd5f188ccc1556663f

### LINKED ISSUE #1713: [firefox][key] Particular keyboard events are not captured by Theia
For instance, pressing <kbd>Ctrl</kbd>+<kbd>-</kbd> is captured in Chrome but not in FF.

On FF:
```
keydown  keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keydown  keyCode=173       which=173       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keypress keyCode=0         which=0         charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=173       which=173       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=false altKey=false metaKey=false
```

On Chrome:
```
keydown  keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keydown  keyCode=189       which=189       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keypress keyCode=31        which=31        charCode=31       
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=189       which=189       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=false altKey=false metaKey=false
```

On Safari:
```
keydown  keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keydown  keyCode=189       which=189       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keypress keyCode=31        which=31        charCode=31       
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=189       which=189       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
keyup    keyCode=17        which=17        charCode=0        
         shiftKey=false ctrlKey=false altKey=false metaKey=false
```

One can see that the <kbd>-</kbd> key was recognized differently on FF with the <kbd>Ctrl</kbd> modifier:
```
keydown  keyCode=173       which=173       charCode=0        
         shiftKey=false ctrlKey=true  altKey=false metaKey=false
```

This blocks #1689.

### LINKED ISSUE #1714: Alt + Right/Left Arrows don't work in the editor
Alt + Right/Left Arrows are usually used to navigate in the editor between words. We override it with the location navigation. We should consider other keybindings for it.

### PR DESCRIPTION
**Pull request by kittaakos**
Closes #1689.

@simark, could you please try it out and verify the functionality? Thanks!
Note: the `last edit location` is not saved and not restored between the sessions.

**Update**:
Before merging it, I need to verify single deletion and insertions/deletion with a range. The insertions/deletion should work with both single and multi-line modifications.

**Update 2**:
Closes #1713 
Closes #1714

I put this PR on hold. -> `WIP`.

### CODE REVIEW

**Comment by kittaakos** [comment]
> the last edit is in a deleted file

We should remove the navigation locations that belong to a deleted resource: https://github.com/theia-ide/theia/issues/1702.

> The ctrl-q shortcut collides with the ctrlcmd+q shortcut

Thanks for checking! This is a blocker.

**Comment by kittaakos** [comment]
Yes. I think that would be sufficient.

Note: https://github.com/theia-ide/theia/blob/aef8737c78bc499ae156e9837895ef7b0d0b4e04/packages/editor/src/browser/editor-navigation-contribution.ts#L47

**Comment by simark** [comment]
Ok.  What's the plan for the shortcut collision?

**Comment by kittaakos** [comment]
> What's the plan for the shortcut collision?

Nothing special, I pick another one. So if you have a recommendation, let me know.

**Comment by kittaakos** [comment]
@simark, this PR changes the keybindings of the `Go Forward` and `Go Back` commands. We try to stick to the VSCode ones. See: https://stackoverflow.com/questions/35424367/how-to-navigate-back-to-the-last-cursor-position-in-visual-studio-code

I have verified it on Windows in Chrome and electron. And on OS X with electron and in both Chrome and Safari.

**Comment by simark** [comment]
> @simark, this PR changes the keybindings of the Go Forward and Go Back commands. We try to stick to the VSCode ones. See: https://stackoverflow.com/questions/35424367/how-to-navigate-back-to-the-last-cursor-position-in-visual-studio-code
> 
> I have verified it on Windows in Chrome and electron. And on OS X with electron and in both Chrome and Safari.

I really like alt-left and alt-right, but consistency with VSCode is probably more important, I'll just have to maintain my personal keybindings file :).

I'll try the PR right now.

**Comment by kittaakos** [comment]
> I really like alt-left and alt-right

Me too.

> consistency with VSCode is probably more important

👍

**Comment by kittaakos** [comment]
> doesn't seem to work on Firefox on Linux

I am running out of ideas, I will check FF on OS X. Thanks for the review!

**Comment by kittaakos** [comment]
> ctrl+shift+- or ctrl+alt+- doesn't seem to work on Firefox on Linux

The <kbd>Ctrl</kbd>+<kbd>-</kbd> does not work on OS X in FF either. So this is a generic problem. I will propose a fix for #1713 in this PR.

**Comment by kittaakos** [comment]
I have verified both `Go Back` and `Go Forward`, and `Go Last Edit Location` on
 - Ubuntu in Chrome and FF.
 - OS X in Chrome, Safari, and FF.
 - Windows in Chrome.

File: packages/core/src/browser/keys.ts [hunk 1]
```diff
@@ -302,26 +302,24 @@ export class KeyCode {
         }
     }
 
-    public static toCode(event: KeyboardEvent): string {
+    public static toCode(event: KeyboardEvent & { readonly keyIdentifier?: string }): string {
```
Comment (simark) [inline packages/core/src/browser/keys.ts]: I wondered where `keyIdentifier` came from.  Perhaps add a comment like:

```
// keyIdentifier is used to access this deprecated field:
// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyIdentifier
```

Comment (kittaakos) [reply packages/core/src/browser/keys.ts]: Yes, you're right. But it was there before as well; we just made an explicit cast to `any`.

Comment (simark) [reply packages/core/src/browser/keys.ts]: Ahh, I see.

**Comment by simark** [comment]
Thanks for putting some info in the commit message by the way!  It's very useful when doing some bisecting/archaeological search.  Next time though, I would suggest wrapping the text to some reasonable (72 or 80 is common).

**Comment by kittaakos** [comment]
> I've just tried it, it works, but there are JS errors in log:

I will double check this.

> Next time though, I would suggest wrapping the text to some reasonable width (72 or 80 is common)

I will do it now.

> I wondered where keyIdentifier came from. Perhaps add a comment like:

👍

**Comment by AlexTugarev** [comment]
@kittaakos if it helps, this is an example for such a key event with error:

screen shot 2018-04-17 at 18 02 36

File: packages/core/src/browser/keys.ts [hunk 2]
```diff
@@ -302,7 +302,15 @@ export class KeyCode {
         }
     }
 
-    public static toCode(event: KeyboardEvent): string {
+    // keyIdentifier is used to access this deprecated field:
+    // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyIdentifier
+    public static toCode(event: KeyboardEvent & { readonly keyIdentifier?: string }): string {
```
Comment (AlexTugarev) [inline packages/core/src/browser/keys.ts]: just asking, why do we need to declare the "extended" type here? especially if that field is deprecated, would there be any other compatible types used on call sites? to access it, we still need to check as it's done at L.323. 
what do I miss?

Comment (kittaakos) [reply packages/core/src/browser/keys.ts]: I wanted to get rid of the `any` cast. That's all.

**Pull request merged**